vortex-datafusion/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+
+        Ok(Self { array, schema_ref })
+    }
+}
+
+#[async_trait]
+impl TableProvider for VortexInMemoryTableProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema_ref)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    /// Plan an array scan.
+    ///
+    /// Currently, no pushdown is supported. The array is flattened directly into the nearest
+    /// Arrow-compatible encoding, and we emit a single [RecordBatch] of data.
+    async fn scan(
+        &self,
+        _state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        _limit: Option<usize>,
+    ) -> DFResult<Arc<dyn ExecutionPlan>> {
+        if !filters.is_empty() {
+            return exec_err!("vortex does not support filter pushdown");
+        }
+
+        let partitioning = if let Ok(chunked_array) = ChunkedArray::try_from(&self.array) {
+            Partitioning::RoundRobinBatch(chunked_array.nchunks())
+        } else {
+            Partitioning::UnknownPartitioning(1)
+        };
+
+        let plan_properties = PlanProperties::new(
+            EquivalenceProperties::new(self.schema_ref.clone()),
+            partitioning,
+            ExecutionMode::Bounded,
+        );
+
+        Ok(Arc::new(VortexMemoryExec {
+            array: self.array.clone(),
+            projection: projection.cloned(),
+            plan_properties,
+        }))
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> DFResult<Vec<TableProviderFilterPushDown>> {
+        // TODO(aduffy): add support for filter pushdown
+        Ok(filters
+            .iter()
+            .map(|_| TableProviderFilterPushDown::Unsupported)
+            .collect())
+    }
+}
+
+/// Physical plan node for scans against an in-memory, possibly chunked Vortex Array.
+#[derive(Debug, Clone)]
+struct VortexMemoryExec {
+    array: Array,
+    projection: Option<Vec<usize>>,
+    plan_properties: PlanProperties,
+}
+
+impl DisplayAs for VortexMemoryExec {
+    fn fmt_as(&self, _display_type: DisplayFormatType, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl VortexMemoryExec {
+    /// Read a single array chunk from the source as a RecordBatch.
+    fn execute_single_chunk(
+        array: Array,
+        projection: &Option<Vec<usize>>,
+        _context: Arc<TaskContext>,
+    ) -> DFResult<SendableRecordBatchStream> {
+        let data = array
+            .flatten()
+            .map_err(|vortex_error| DataFusionError::Execution(format!("{}", vortex_error)))?
+            .into_array();
+
+        // Construct the RecordBatch by flattening each struct field and transmuting to an ArrayRef.
+        let struct_array = StructArray::try_from(data)
+            .map_err(|_| exec_datafusion_err!("top-level array must be Struct encoding"))?;
```

Comment:
Since this is already checked, can you `.expect(...)`?

Reply:
Sure. I'm still a bit fuzzy where to the draw line on panicking VS returning an error for functions that are already returning Result type

Reply:
You do the check to be able to return the error and at that point if you need to unwrap the value you can panic since it should never happen (tm).

Reply:
done